src/utils/buildMetadata.ts

```diff
@@ -166,13 +153,184 @@ export function buildMetadata(sheets: Sheet[], defaultCC: string) {
         trackedEntityTypes,
-        //programs,
+        programSections: buildprogramSections(sheets),
+        programs: buildPrograms(sheets),
+        programStages: buildProgramStages(sheets),
+        programStageSections: buildProgramStageSections(sheets),
         programRules: buildProgramRules(sheets),
         programRuleActions: buildProgramRuleActions(sheets),
         programRuleVariables: buildProgramRuleVariables(sheets),
     };
 }
 
+function buildPrograms(sheets: Sheet[]) {
+    const get = (name: string) => getItems(sheets, name);
+
+    const programs = get("programs");
+    const pSections = get("programSections");
+    const pStages = get("programStages");
+    const trackedEntityTypes = get("trackedEntityTypes");
+    const categoryCombos = get("categoryCombos");
+
+    return programs.map(program => {
+        let data = { ...program } as MetadataItem;
+
+        const trackedEntityType = {
+            id: getByName(trackedEntityTypes, program.trackedEntityType)?.id
+        };
+
+        const programStages = pStages.filter((pStageToFilter) => {
+            return pStageToFilter.program === program.name;
+        }).map(programStage => ({ id: programStage.id }));
+
+        replaceById(data, "categoryCombo", categoryCombos);
+
+        if (trackedEntityType.id) {
+            // WITH_REGISTRATION == Tracker Program
+            const programType = "WITH_REGISTRATION";
+
+            // Event Program Stages belong to programStageSections
+            const programSections = pSections.filter((pSectionToFilter) => {
+                return pSectionToFilter?.program === program.name;
+            }).map(programSection => ({ id: programSection.id }));
+
+            return { ...data, programType, trackedEntityType, programStages, programSections };
+        } else {
+            // WITHOUT_REGISTRATION == Event Program
+            const programType = "WITHOUT_REGISTRATION";
+            return { ...data, programType, programStages };
+        }
+    });
+}
+
+function buildprogramSections(sheets: Sheet[]) {
+    const get = (name: string) => getItems(sheets, name);
+
+    const programSections = _.cloneDeep(get("programSections"));
+    const programs = get("programs");
+    const teAttributes = get("trackedEntityAttributes");
+
+    const sectionsAttributes = get("programSectionsTrackedEntityAttributes").map(psTrackedEntityAttribute => {
+        const programSection = programSections.find(pSectionToFind => {
+            return pSectionToFind.name === psTrackedEntityAttribute.programSection &&
+                pSectionToFind.program === psTrackedEntityAttribute.program
+        })?.id;
+
+        const trackedEntityAttribute = getByName(teAttributes, psTrackedEntityAttribute.name)?.id;
+
+        return { programSection, trackedEntityAttribute }
+    });
+
+    return programSections.map(programSection => {
+        const program = {
+            id: getByName(programs, programSection.program)?.id
+        };
```

Comment:
A simpler way, that also doesn't need to add `program` in the `return { ...programSection, program, [...]}` line, is to do:

```
replaceById(programSection, "program", programs);
```

(You can see how replaceById() is used in other functions, like buildProgramRuleVariables() and more - this is such a common operation that seems worth doing it always in the same way!)

Reply:
We need the name of the program associated to the Section, so I kept `programSection.program` for that.
I could store the program name in a new variable and use `replaceById()` if its more readable.

Reply:
Uhm, `replaceById(programSection, "program", programs)` would already use `programSection.program` (in the form of `data[key]`, with `data = programSection` and `key = "program"` in line 428). So I think you can use the `replaceById(...)` thing directly?

But I'm not sure I'm understanding correctly...

Reply:
What i mean is that we need the original value of `programSection.program` for the next code chunk:
```
if (programSection.sortOrder === undefined) {
    addSortOrder(programSections.filter((pSectionToFilter) => {
        return pSectionToFilter.program === programSection.program;
    }));
}
```

So how about i just swap them in order to use `replaceById`?
```
if (programSection.sortOrder === undefined) {
    addSortOrder(programSections.filter((pSectionToFilter) => {
        return pSectionToFilter.program === programSection.program;
    }));
}

replaceById(programSection, "program", programs)
```